cpp/louvain_module/algorithm/louvain.cpp

```diff
@@ -0,0 +1,145 @@
+#include "louvain.hpp"
+
+namespace {
+
+constexpr int kReplaceMap = 0;
+constexpr int kThreadsOpt = 1;
+constexpr int kNumColors = 16;
+
+std::vector<std::uint64_t> GrappoloCommunityDetection(graph *grappolo_graph, bool coloring,
+                                                      std::uint64_t min_graph_size, double threshold,
+                                                      double coloring_threshold) {
+  auto number_of_vertices = grappolo_graph->numVertices;
+
+  int num_threads = 1;  // Default is one thread
+#pragma omp parallel
+  { num_threads = omp_get_num_threads(); }
+
+  // Initialize clustering array
+  auto *cluster_array = (long *)malloc(number_of_vertices * sizeof(long));
+#pragma omp parallel for
+  for (long i = 0; i < number_of_vertices; i++) {
+    cluster_array[i] = -1;
+  }
+
+  if (coloring) {
+    runMultiPhaseColoring(grappolo_graph, cluster_array, coloring, kNumColors, kReplaceMap, min_graph_size, threshold,
+                          coloring_threshold, num_threads, kThreadsOpt);
+  } else {
+    runMultiPhaseBasic(grappolo_graph, cluster_array, kReplaceMap, min_graph_size, threshold, coloring_threshold,
+                       num_threads, kThreadsOpt);
+  }
+
+  // Store clustering information in vector
+  std::vector<std::uint64_t> result;
+  for (long i = 0; i < number_of_vertices; ++i) {
+    result.emplace_back(cluster_array[i]);
+  }
+  // Detach memory, no need to free graph instance, it will be removed inside algorithm
+  free(cluster_array);
+
+  // Return empty vector if algorithm has failed
+  return result;
+}
+
+void LoadUndirectedEdges(const mg_graph::GraphView<> &memgraph_graph, graph *grappolo_graph) {
+  int num_threads = 1;
+#pragma omp parallel
+  { num_threads = omp_get_num_threads(); }
+
+  auto number_of_vertices = memgraph_graph.Nodes().size();
+  auto number_of_edges = memgraph_graph.Edges().size();
+
+  // Case without graph edges
+  if (number_of_edges == 0) {
+    grappolo_graph->numEdges = 0;
+    return;
+  }
+
+  auto tmp_edge_list = std::unique_ptr<edge[]>(new edge[number_of_edges]);  // Every edge stored ONCE
```

Comment:
I am mostly concerned with introducing yet another implementation of graph, edge from grappolo library, and it doesn't even have a namespace... Maybe it's not an issue now, what do you think @antonio2368

Reply:
I agree, this is getting a bit confusing, especially without a namespace, but not sure how to solve this issue...